protobuf/protoc-gen-cruxclient/cruxclient_generator.cc

```diff
@@ -0,0 +1,426 @@
+    } else {
+      static char hex[] = "0123456789abcdef";
+      result.push_back('_');
+      result.push_back(hex[(c >> 4) & 0xf]);
+      result.push_back(hex[c & 0xf]);
+    }
+  }
+  return result;
+}
+
+inline string StringReplace(string str, const string &from, const string &to,
+                            bool replace_all) {
+  size_t pos = 0;
+
+  do {
+    pos = str.find(from, pos);
+    if (pos == string::npos) {
+      break;
+    }
+    str.replace(pos, from.length(), to);
+    pos += to.length();
+  } while (replace_all);
+
+  return str;
+}
+
+inline string StringReplace(string str, const string &from, const string &to) {
+  return StringReplace(str, from, to, true);
+}
+
+// The following 4 function was copied from
+// grpc/src/compiler/generator_helpers.h
+inline std::vector<string> tokenize(const string &input,
+                                    const string &delimiters) {
+  std::vector<string> tokens;
+  size_t pos, last_pos = 0;
+
+  for (;;) {
+    bool done = false;
+    pos = input.find_first_of(delimiters, last_pos);
+    if (pos == string::npos) {
+      done = true;
+      pos = input.length();
+    }
+
+    tokens.push_back(input.substr(last_pos, pos - last_pos));
+    if (done) return tokens;
+
+    last_pos = pos + 1;
+  }
+}
+
+// The following 3 function was copied from
+// grpc/src/compiler/cpp_generator_helpers.h
+
+inline string DotsToColons(const string &name) {
+  return StringReplace(name, ".", "::");
+}
+
+inline string DotsToUnderscores(const string &name) {
+  return StringReplace(name, ".", "_");
+}
+
+inline string ClassName(const Descriptor *descriptor, bool qualified) {
+  // Find "outer", the descriptor of the top-level message in which
+  // "descriptor" is embedded.
+  const Descriptor *outer = descriptor;
+  while (outer->containing_type() != NULL) outer = outer->containing_type();
+
+  const string &outer_name = outer->full_name();
+  string inner_name = descriptor->full_name().substr(outer_name.size());
+
+  if (qualified) {
+    return DotsToColons(outer_name) + DotsToUnderscores(inner_name);
+  } else {
+    return outer->name() + DotsToUnderscores(inner_name);
+  }
+}
+
+void PrintPrologue(Printer *printer, const FileDescriptor *file) {
+  std::map<string, string> vars;
+  vars["filename"] = file->name();
+  printer->Print("// Generated by the CRUX Service Client C++ plugin.\n");
+  printer->Print("// If you make any local change, they will be lost.\n");
+  printer->Print(vars, "// source: $filename$\n");
+}
+
+void PrintNamespace(Printer *printer, const FileDescriptor *file,
+                    bool isEpilogue) {
+  std::map<string, string> vars;
+  if (!file->package().empty()) {
+    std::vector<string> parts = tokenize(StripProto(file->package()), ".");
+    for (auto part = parts.begin(); part != parts.end(); part++) {
+      vars["part"] = *part;
+      if (!isEpilogue) {
+        printer->Print(vars, "namespace $part$ {\n");
+      } else {
+        printer->Print(vars, "}\n");
```

Comment:
Linting will be ignored because this will be auto-generated code.
I originally had this printed out, however this requires extra logic to print the namespaces closing in reverse order; so wasn't worth the extra complexity, as this is not meant for human readable code.